File: Project-Euler/Solutions-Simon/euler47/euler47.jl
#=
The first two consecutive numbers to have two distinct prime factors are:

14 = 2 × 7
15 = 3 × 5

The first three consecutive numbers to have three distinct prime factors are:

644 = 2² × 7 × 23
645 = 3 × 5 × 43
646 = 2 × 17 × 19.

Find the first four consecutive integers to have four distinct prime factors. What is the first of these numbers?
=#
#Answer:

#Terminal: Julia euler47.jl 4 1000000
cons = parse(Int, ARGS[1])
limit = parse(Int, ARGS[2])

function numFactors(number, primeList)
  nod = 0
  pf = false
  remain = number

  for i = 1 : length(primeList)
    if primeList[i] * primeList[i] > number
      nod += 1
      return nod
    end

    pf = false
    while rem(remain, primeList[i]) == 0
      pf = true
      remain = div(remain, primeList[i])
    end
    if pf
      nod += 1
    end

    if remain == 1
      return nod
    end
  end
  return nod
end

function getPrimes(size)
	array = fill(true, size)
	c = 1
	for i = 2 : size
		if array[i] == true
			j = i + i
			while j <= size
				array[j] = false
				j += i
			end
			c += 1
		end
	end
	primeArray = Int[]
	for i = 2 : size
		if(array[i])
			push!(primeArray, i)
		end
	end
	return primeArray
end

function calc(limit, consec)
  primeList = getPrimes(limit)
  targetpf = consec
  targetcons = consec
  cons = 1
  result = 1

  while cons < targetcons
    result += 1
    if numFactors(result, primeList) >= targetpf
      cons += 1
    else
      cons = 0
    end
  end
  return result-3
end

println(calc(limit, cons))

File: Project-Euler/Stats/euler47/euler47.cpp
//
//  main.cpp
//  euler47
//
//  Created by Sonni on 23/05/16.
//  Copyright © 2016 Sonni Jensen. All rights reserved.
//

#include <iostream>
#include <vector>

unsigned int numFactors(unsigned int number, std::vector<unsigned int>& primeList)
{
    unsigned int nod = 0;
    bool pf = false;
    unsigned int remain = number;
    
    for (unsigned int i = 0; i < primeList.size(); i++)
    {
        if (primeList[i] * primeList[i] > number)
        {
            nod++;
            return nod;
        }
        
        pf = false;
        while (remain % primeList[i] == 0)
        {
            pf = true;
            remain = remain / primeList[i];
        }
        if (pf)
            nod++;
        
        if (remain == 1)
            return nod;
    }
    return nod;
}

std::vector<unsigned int> getPrimes(unsigned int size)
{
    std::vector<bool> array;
    for (unsigned int i = 0; i <= size; i++)
        array.push_back(true);

    for (unsigned int i = 2; i <= size; i++)
    {
        if (array[i])
        {
            unsigned int j = i + i;
            while (j <= size)
            {
                array[j] = false;
                j += i;
            }
        }
    }
    std::vector<unsigned int> primeArray;
    for (unsigned int i = 2; i <= size; i++)
        if (array[i])
            primeArray.push_back(i);
    
    return primeArray;
}


unsigned int calc(unsigned int limit, unsigned int consec)
{
    std::vector<unsigned int> primeList = getPrimes(limit);
    unsigned int targetpf = consec;
    unsigned int targetcons = consec;
    unsigned int cons = 1;
    unsigned int result = 1;
    
    while (cons < targetcons)
    {
        result++;
        if (numFactors(result, primeList) >= targetpf)
            cons++;
        else
            cons = 0;
    }
    
    return result-3;
}

int main(int argc, const char * argv[]) {
    // insert code here...
    std::cout << calc(atoi(argv[2]), atoi(argv[1])) << std::endl;
    return 0;
}
